File: Association Rule Learning/Section 28 - Apriori/apriori.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Fri Feb 28 18:24:12 2020

@author: kumet
"""

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

dataset = pd.read_csv("Market_Basket_Optimisation.csv", header = None)
transactions = []
for i in range(0,7501):
    transactions.append([str(dataset.values[i,j]) for j in range(0,20)])
    
#training Apriori on the dataset
from apyori import apriori
support = 3*7/7500 #support of a product that is purchased 3 times a day
rules = apriori(transactions,min_support = support,min_confidence = 0.2, min_lift = 3, min_length = 2)

#visualising the results
results = list(rules) #their relavence is sorted by a combination of support, lift, confidence by the apyori class.

File: Association Rule Learning/Section 28 - Apriori/apriori.R
#Apriori

# data preprocessing
library(arules)
#dataset = read.csv("Market_Basket_Optimisation.csv", header = FALSE)
dataset = read.transactions("Market_Basket_Optimisation.csv", sep = ",",rm.duplicates = TRUE)
#>distribution of transactions with duplicates:
#  1 
#  5
# 5 transactions including 1 duplicate
summary(dataset)
itemFrequencyPlot(dataset,topN = 100)

#support = 3*7/7500=0.0028 = 0.003
#support = 4*7/7500=0.0037 = 0.004
#training Apriori on the dataset
rules = apriori(data = dataset, parameter = list(support = 0.004,confidence = 0.2 ))
#confidence is 0.8 by default

inspect(sort(rules, by = "lift")[1:10])
#this code makes sense but burned my brain 